executor/executor.go

```diff
@@ -2348,3 +2362,457 @@ func isWeakConsistencyRead(ctx sessionctx.Context, node ast.Node) bool {
+			return i.bucket < j.bucket
+		})
+
+		currentOffset := 0
+
+		// Every checksum in table side should be the same as the index side.
+		i := 0
+		for i < len(tableChecksum) && i < len(indexChecksum) {
+			if tableChecksum[i].bucket != indexChecksum[i].bucket || tableChecksum[i].checksum != indexChecksum[i].checksum {
+				if tableChecksum[i].bucket <= indexChecksum[i].bucket {
+					currentOffset = int(tableChecksum[i].bucket)
+					tableRowCntToCheck = tableChecksum[i].count
+				} else {
+					currentOffset = int(indexChecksum[i].bucket)
+					tableRowCntToCheck = indexChecksum[i].count
+				}
+				meetError = true
+				break
+			}
+			i++
+		}
+
+		if !meetError && i < len(indexChecksum) && i == len(tableChecksum) {
+			// Table side has fewer buckets.
+			currentOffset = int(indexChecksum[i].bucket)
+			tableRowCntToCheck = indexChecksum[i].count
+			meetError = true
+		} else if !meetError && i < len(tableChecksum) && i == len(indexChecksum) {
+			// Index side has fewer buckets.
+			currentOffset = int(tableChecksum[i].bucket)
+			tableRowCntToCheck = tableChecksum[i].count
+			meetError = true
+		}
+
+		if !meetError {
+			if times != 1 {
+				logutil.BgLogger().Error("unexpected result, no error detected in this round, but an error is detected in the previous round", zap.Int("times", times), zap.Int("offset", offset), zap.Int("mod", mod))
+			}
+			break
+		}
+
+		offset += currentOffset * mod
+		mod *= bucketSize
+	}
+
+	queryToRow := func(se sessionctx.Context, sql string) ([]chunk.Row, error) {
+		rs, err := se.(sqlexec.SQLExecutor).ExecuteInternal(ctx, sql)
+		if err != nil {
+			return nil, err
+		}
+		row, err := sqlexec.DrainRecordSet(ctx, rs, 4096)
+		if err != nil {
+			return nil, err
+		}
+		err = rs.Close()
+		if err != nil {
+			logutil.BgLogger().Warn("close result set failed", zap.Error(err))
+		}
+		return row, nil
+	}
+
+	if meetError {
+		groupByKey := fmt.Sprintf("((%s - %d) %% %d)", md5Handle.String(), offset, mod)
+		indexSQL := fmt.Sprintf("select %s, %s, %s from %s use index(`%s`) where %s = 0 order by %s", handleColumnField, indexColumnField.String(), md5HandleAndIndexCol.String(), dbutil.TableName(w.e.dbName, w.e.table.Meta().Name.String()), idxInfo.Name, groupByKey, handleColumnField)
+		tableSQL := fmt.Sprintf("select /*+ read_from_storage(tikv[%s]) */ %s, %s, %s from %s use index() where %s = 0 order by %s", dbutil.TableName(w.e.dbName, w.e.table.Meta().Name.String()), handleColumnField, indexColumnField.String(), md5HandleAndIndexCol.String(), dbutil.TableName(w.e.dbName, w.e.table.Meta().Name.String()), groupByKey, handleColumnField)
+
+		idxRow, err := queryToRow(se, indexSQL)
+		if err != nil {
+			trySaveErr(err)
+			return
+		}
+		tblRow, err := queryToRow(se, tableSQL)
+		if err != nil {
+			trySaveErr(err)
+			return
+		}
+
+		getHandleFromRow := func(row chunk.Row) (kv.Handle, error) {
+			handleDatum := make([]types.Datum, 0)
+			for i, t := range pkTypes {
```

Comment:
`pkTypes` is always nil? I only see it is set in https://github.com/pingcap/tidb/pull/43694/files#diff-4c9a4c1e9a402d07a0a78f4db9f766d23f1ff57ceda1e75c4e83b579488ad301R2460.